### Importing libraries to create the dashboard

In [1]:
# Import libraries
from dotenv import load_dotenv
import os
import psycopg2
from sqlalchemy import create_engine, MetaData
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import ipywidgets as widgets
from IPython.display import display

### Load Environment Variables

In [2]:
# Load environment variables
load_dotenv()
db_user = os.getenv('DB_USER')
db_password = os.getenv('DB_PASSWORD')
db_host = os.getenv('DB_HOST')
db_name = os.getenv('DB_NAME')

### Connect to the Online Database

In [3]:
# Connect to the Stack Overflow database
conn = psycopg2.connect(
    host=db_host,
    database=db_name,
    user=db_user,
    password=db_password
)

### Defining the Heart Metrics and storing it in a Dictionary

In [4]:
# Define HEART metrics for Stack Overflow
heart_metrics = {
    'Happiness': 'Measure of user satisfaction and happiness with the Questions & Answers product.',
    'Engagement': 'Measure of user interaction and activity on the platform.',
    'Adoption': 'Measure of user acquisition and growth of Stack Overflow user base.',
    'Retention': 'Measure of user retention and continued usage of the platform over time.',
    'Task Success': 'Measure of user ability to accomplish their goals and tasks effectively on Stack Overflow.'
}

### Creating a SQLAlchemy Engine to be used for Data Exploration with Pandas

In [5]:
# Enable sql extension
%load_ext sql

In [6]:
engine = create_engine(f"postgresql+psycopg2://{db_user}:{db_password}@{db_host}/{db_name}")
connection = engine.connect()

In [7]:
# Check what tables exists within the database
metadata = MetaData()
metadata.reflect(bind=engine)

tables = metadata.tables
tables_list = [table for table in tables]
print("Tables in the database:")
tables_list

Tables in the database:


['users', 'posts', 'votes', 'tags', 'comments']

In [8]:
# Test the connection through users table
test_query = "SELECT * FROM users LIMIT 5"
test_df = pd.read_sql(test_query, connection)
test_df

id  account_id  reputation  views  down_votes  up_votes  \
0  13990332    19152802           1      1           0         0   
1  13990333    19152803           1      1           0         0   
2  13990334    19152804           1      2           0         0   
3  13990335    19152805           1      1           0         0   
4  13990336    19152807           1      5           0         0   

         display_name location profile_image_url website_url about_me  \
0            dropkicc     None              None        None     None   
1     Carlos Calderon     None              None        None     None   
2       Yogesh Mathur    India              None        None     None   
3       ปฐมพร เมฆทวีป     None              None        None     None   
4  LuciferMorningstar     None              None        None     None   

            creation_date        last_access_date  
0 2020-07-24 18:26:03.427 2020-12-31 01:53:23.433  
1 2020-07-24 18:26:16.733 2022-10-16 00:21:02.993  
2 2020-07-24 18:26:22.877 2020-08-16 15:38:01.953  
3 2020-07-24 18:26:32.657 2020-07-24 20:27:53.290  
4 2020-07-24 18:26:34.047 2022-03-17 17:32:53.177

### Data Exploration on Tables
After having a look at the notes made in my plan.md, I think understand what that `Posts` table would be good to explore first. This is because I don't really know what posts users can be making and getting an understanding of this first would give me a better of what the product is for the users.

After getting a grasp of what the product is, looking at the votes and feedback could be a good next step.

In [9]:
# Exploring Posts table
posts_query = "SELECT * FROM Posts ORDER BY id LIMIT 5"
posts_df = pd.read_sql(posts_query, connection)
posts_df

id  owner_user_id  last_editor_user_id  post_type_id  accepted_answer_id  \
0   4              8             16124033             1                 7.0   
1   6              9              9134576             1                31.0   
2   7              9              5496973             2                 NaN   
3   9              1              3524942             1              1404.0   
4  11              1             16790137             1              1248.0   

   score  parent_id  view_count  answer_count  comment_count  ...  \
0    787        NaN     69172.0          14.0              7  ...   
1    319        NaN     23239.0           7.0              0  ...   
2    518        4.0         NaN           NaN              1  ...   
3   2179        NaN    770310.0          71.0             10  ...   
4   1636        NaN    195320.0          41.0              3  ...   

                                               title  \
0            How to convert Decimal to Double in C#?   
1  Why did the width collapse in the percentage w...   
2                                               None   
3  How do I calculate someone's age based on a Da...   
4                      Calculate relative time in C#   

                                                tags content_license  \
0  <c#><floating-point><type-conversion><double><...    CC BY-SA 4.0   
1                   <html><css><internet-explorer-7>    CC BY-SA 4.0   
2                                               None    CC BY-SA 4.0   
3                               <c#><.net><datetime>    CC BY-SA 4.0   
4  <c#><datetime><time><datediff><relative-time-s...    CC BY-SA 4.0   

                                                body favorite_count  \
0  <p>I want to assign the decimal variable &quot...            0.0   
1  <p>I have an absolutely positioned <code>div</...            0.0   
2  <p>An explicit cast to <code>double</code> lik...            NaN   
3  <p>Given a <code>DateTime</code> representing ...            0.0   
4  <p>Given a specific <code>DateTime</code> valu...            0.0   

            creation_date    community_owned_date closed_date  \
0 2008-07-31 21:42:52.667 2012-10-31 16:42:47.213        None   
1 2008-07-31 22:08:08.620                     NaT        None   
2 2008-07-31 22:17:57.883                     NaT        None   
3 2008-07-31 23:40:59.743 2011-08-16 19:40:43.080        None   
4 2008-07-31 23:55:37.967 2009-09-04 13:15:59.820        None   

           last_edit_date      last_activity_date  
0 2022-09-08 05:07:26.033 2022-09-08 05:07:26.033  
1 2021-01-29 18:46:45.963 2021-01-29 18:46:45.963  
2 2019-10-21 14:03:54.607 2019-10-21 14:03:54.607  
3 2022-07-27 22:34:36.320 2022-11-29 15:05:54.720  
4 2022-07-10 00:19:55.237 2022-09-05 11:26:30.187  

[5 rows x 22 columns]

This is really interesting and gave me insights as to what this table is about. From the rows here, I can see there are a few other tables it's connecting up to i.e. PostTypes, Users, PostNotices. As there are 22 columns in total, I want to select a few that are numeric i.e. comment and answer count to understand better what these values can mean.

In [10]:
posts_df.columns

Index(['id', 'owner_user_id', 'last_editor_user_id', 'post_type_id',
       'accepted_answer_id', 'score', 'parent_id', 'view_count',
       'answer_count', 'comment_count', 'owner_display_name',
       'last_editor_display_name', 'title', 'tags', 'content_license', 'body',
       'favorite_count', 'creation_date', 'community_owned_date',
       'closed_date', 'last_edit_date', 'last_activity_date'],
      dtype='object')

In [11]:
posts_df_filtered = posts_df[['id', 'score', 'view_count', 'answer_count', 'comment_count','favorite_count']]
posts_df_filtered

id  score  view_count  answer_count  comment_count  favorite_count
0   4    787     69172.0          14.0              7             0.0
1   6    319     23239.0           7.0              0             0.0
2   7    518         NaN           NaN              1             NaN
3   9   2179    770310.0          71.0             10             0.0
4  11   1636    195320.0          41.0              3             0.0

The rows here are particularly helpful, giving an overview what these numbers mean, and we can also see that some rows are missing data or have 0 counts. 

In [12]:
# Check what table names are from tables list
tables_list

['users', 'posts', 'votes', 'tags', 'comments']

### Exploring the Votes and Feedback data

In [13]:
votes_query = "SELECT * FROM votes ORDER BY id LIMIT 5"
votes_df = pd.read_sql(votes_query, connection)
votes_df

id user_id  post_id  vote_type_id bounty_amount creation_date
0   1    None        1             2          None    2008-07-31
1   2    None        3             2          None    2008-07-31
2   3    None        2             2          None    2008-07-31
3   4    None        4             2          None    2008-07-31
4   5    None        6             2          None    2008-07-31

This looks interesting and as a next step, I want to see what the unique values are for the vote_type_id so I am going to have grab more rows to find out about this. For now, I can see that a user id and a post id is link to each vote and also there is an interesting column bounty_amount - would love to understand what this is about!

In [14]:
# Let's grab the first 100 rows
votes_query = "SELECT * FROM votes ORDER BY id LIMIT 100"
votes_df = pd.read_sql(votes_query, connection)
votes_df

id user_id  post_id  vote_type_id bounty_amount creation_date
0     1    None        1             2          None    2008-07-31
1     2    None        3             2          None    2008-07-31
2     3    None        2             2          None    2008-07-31
3     4    None        4             2          None    2008-07-31
4     5    None        6             2          None    2008-07-31
..  ...     ...      ...           ...           ...           ...
95  194    None      127             2          None    2008-08-01
96  195    None      126             2          None    2008-08-01
97  200    None      117             3          None    2008-08-01
98  201    None      111             2          None    2008-08-01
99  207    None      127             1          None    2008-08-01

[100 rows x 6 columns]

In [15]:
# Find out the unique values of vote_type_id
votes_df['vote_type_id'].unique()

array([2, 3, 1])

Okay so for now there seems to be at least 3 different values for vote types meaning there are at least 3 different types of votes you can do on Stackoverflow 

In [16]:
# Let's get some info on the votes columns and it's data types
votes_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   id             100 non-null    int64         
 1   user_id        0 non-null      object        
 2   post_id        100 non-null    int64         
 3   vote_type_id   100 non-null    int64         
 4   bounty_amount  0 non-null      object        
 5   creation_date  100 non-null    datetime64[ns]
dtypes: datetime64[ns](1), int64(3), object(2)
memory usage: 4.8+ KB


Okay so I wanted to check the bounty_amount column to see if there are any non null values and they all are. Let's see if I can grab some non-null rows and save it in a new dataframe

In [24]:
votes_query = "SELECT * FROM votes WHERE bounty_amount IS NOT NULL ORDER BY id LIMIT 5"
votes_df = pd.read_sql(votes_query, connection)
votes_df

id  user_id  post_id  vote_type_id  bounty_amount creation_date
0  1472821     4234    98225             8            350    2009-01-27
1  1472847     4234   296603             8            550    2009-01-27
2  1472968     2140   309711             8            250    2009-01-27
3  1473097     2635   222716             8            150    2009-01-27
4  1473113     2822    74883             8            200    2009-01-27

This is interesting to see as we can see there can be a bounty amount for votes. Not sure what this really means. Last thing I want to explore before we get too deep is how many non null values rows there are for bounty_amount

In [25]:
votes_query = "SELECT * FROM votes WHERE bounty_amount IS NOT NULL ORDER BY id"
votes_df = pd.read_sql(votes_query, connection)
votes_df

id    user_id   post_id  vote_type_id  bounty_amount  \
0         1472821     4234.0     98225             8            350   
1         1472847     4234.0    296603             8            550   
2         1472968     2140.0    309711             8            250   
3         1473097     2635.0    222716             8            150   
4         1473113     2822.0     74883             8            200   
...           ...        ...       ...           ...            ...   
448909  265133236  8165583.0  75612754             8             50   
448910  265133583   282328.0  75590032             8            100   
448911  265134182  1103606.0  75573903             8             50   
448912  265135287  5328691.0  75626337             8             50   
448913  265136301        NaN  75614199             9            200   

       creation_date  
0         2009-01-27  
1         2009-01-27  
2         2009-01-27  
3         2009-01-27  
4         2009-01-27  
...              ...  
448909    2023-03-05  
448910    2023-03-05  
448911    2023-03-05  
448912    2023-03-05  
448913    2023-03-05  

[448914 rows x 6 columns]

Okay this is a lot of rows (44k) with bounty_amount, also shows this dataset is huge so need to be careful when making visualisations and analysis

### Data Exploration on Comments table

In [27]:
comments_query = "SELECT * FROM comments LIMIT 5"
comments_df = pd.read_sql(comments_query, connection)
comments_df

id  post_id  user_id  score content_license user_display_name  \
0  10    45651      242      6    CC BY-SA 2.5              None   
1  12    47428     4642      3    CC BY-SA 2.5              None   
2  14    47481     4642      0    CC BY-SA 2.5              None   
3  15    47373     2495      0    CC BY-SA 2.5              None   
4  16    47497     4642      1    CC BY-SA 2.5              None   

                                                text           creation_date  
0  It will help if you give some details of which... 2008-09-06 13:38:23.647  
1  One of the things that make a url user-friendl... 2008-09-06 13:51:47.843  
2  I agree, both CodeRush and RefactorPro are vis... 2008-09-06 14:15:46.897  
3  Just wanted to mention that this is an excelle... 2008-09-06 14:30:40.217  
4  Indeed, the only way to do this is get the ser... 2008-09-06 14:42:35.303

Okay from this we can see it's got some interesting details, particularly the score where this is giving us an idea how popular or helpful this comment is for other users. 

In [28]:
# Let's do a quick check on the data types of these columns
comments_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   id                 5 non-null      int64         
 1   post_id            5 non-null      int64         
 2   user_id            5 non-null      int64         
 3   score              5 non-null      int64         
 4   content_license    5 non-null      object        
 5   user_display_name  0 non-null      object        
 6   text               5 non-null      object        
 7   creation_date      5 non-null      datetime64[ns]
dtypes: datetime64[ns](1), int64(4), object(3)
memory usage: 452.0+ bytes


Nothing unusual here, good to know that the creation date is a datetime 

### Data Exploration on tags table

In [29]:
tables_list

['users', 'posts', 'votes', 'tags', 'comments']

In [31]:
tags_query = "SELECT * FROM tags ORDER BY id LIMIT 5"
tags_df = pd.read_sql(tags_query, connection)
tags_df

id  excerpt_post_id  wiki_post_id    tag_name    count
0   1          3624959       3607476        .net   329455
1   2          3673183       3673182        html  1167742
2   3          3624960       3607052  javascript  2479947
3   4          3644670       3644669         css   787138
4   5          3624936       3607050         php  1456271

Okay this is giving me a bit more insights as to what this table is showing. I believe the count column is counting how many posts have that tag_name i.e. .net has 329455 posts. excerpt_post_id and wiki_post_id are the ids associated with the post and the wiki post using that tag.

### Data Exploration on the users table

In [32]:
users_query = "SELECT * FROM users ORDER BY id LIMIT 5"
users_df = pd.read_sql(users_query, connection)
users_df

id account_id  reputation  views  down_votes  up_votes     display_name  \
0 -1014       None           1      0           0         0       R Language   
1 -1013       None           1      0           0         0            CI/CD   
2 -1012       None           1      0           0         0            Audio   
3 -1011       None           1      0           0         0              AWS   
4 -1010       None           1      0           0         0  Microsoft Azure   

  location profile_image_url website_url  \
0     None              None        None   
1     None              None        None   
2     None              None        None   
3     None              None        None   
4     None              None        None   

                                            about_me           creation_date  \
0  <p>A collective where data scientists and AI r... 2023-02-17 19:52:56.213   
1  <p>A collective where developers focused on co... 2023-02-17 19:25:19.953   
2  <p>Official Q&amp;A for Audio support. This co... 2023-02-15 23:24:13.363   
3  <p>Amazon Web Services (AWS) is the world’s mo... 2022-11-09 20:49:03.637   
4  <p>On-premises, hybrid, multicloud, or at the ... 2022-10-25 19:18:31.537   

         last_access_date  
0 2023-02-17 19:52:56.280  
1 2023-02-17 19:25:20.037  
2 2023-02-15 23:24:13.373  
3 2022-11-09 20:49:03.637  
4 2022-10-25 19:18:31.537

Once of the interesting things I have noticed is that there are negative ids here. I am not sure if it is because stackoverflow has so many users that it ran out of positive numbers or something bizzare happened one day but something imortant to note. 

Otherwise this table looks pretty standard in terms of the fact it's storing user profile information i.e. display name, how people have viewed their profile, their bio, when they were created and when they last used stack overflow. I think when going in and creating the dashboard, the important metrics to look at from this table will be their creation date and last access date as well as reputation, views, down_votes and up_votes to see how engaged they are in using the product